### PR TITLE
Refactor so other DI Systems can extend eureka.

### PR DESCRIPTION
**Pull request by spencergibb**
Make EurekaJacksonCodec more inline with Converters with public static classes to extend.
Make EurekaJacksonCodec.INSTANCE mutable.
Add get/set to DiscoveryClientOptionalArgs.

I'd rather not have a static instance in EurekaJacksonCodec, suggestions welcome.

### CODE REVIEW

**Comment by spencergibb** [comment]
Not sure of the build failure.  Didn't change any behavior.

File: eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java [hunk 1]
```diff
@@ -196,6 +196,30 @@
 
         @Inject(optional = true)
         private Provider<HealthCheckHandler> healthCheckHandlerProvider;
+
+        public EventBus getEventBus() {
```
Comment (tbak) [inline eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java]: I understand that these methods are an alternative to DiscoveryClientOptionalArgs. As we use constructor injection everywhere, why not add yet another constructor that accepts the additional, optional values directly without a wrapper?

File: eureka-client/src/main/java/com/netflix/discovery/converters/EurekaJacksonCodec.java [hunk 2]
```diff
@@ -50,7 +50,7 @@
 
     private static final Logger logger = LoggerFactory.getLogger(EurekaJacksonCodec.class);
 
-    private static final Version VERSION = new Version(1, 1, 0, null);
+    protected static final Version VERSION = new Version(1, 1, 0, null);
```
Comment (tbak) [inline eureka-client/src/main/java/com/netflix/discovery/converters/EurekaJacksonCodec.java]: Is this refactoring needed to be able to extend serializer to support other data center types?

File: eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java (hunk 1 above)
Comment (spencergibb) [inline eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java]: They aren't an "alternative to DiscoveryClientOptionalArgs" as they are methods on the class.  I'll add the new constructor instead.

File: eureka-client/src/main/java/com/netflix/discovery/converters/EurekaJacksonCodec.java (hunk 2 above)
Comment (spencergibb) [inline eureka-client/src/main/java/com/netflix/discovery/converters/EurekaJacksonCodec.java]: probably not.

File: eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java (hunk 1 above)
Comment (spencergibb) [inline eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java]: Hmm, adding a new constructor to `DiscoveryClient` would seem more invasive than needs be.  Let me update and see what you think.

**Comment by spencergibb** [comment]
@tbak I make the `DiscoveryClientOptionalArgs` have a constructor.  It seems to be the simplest thing to do what I'd like to accomplish.

**Comment by spencergibb** [comment]
@tbak ping, I've rebased against master.

**Comment by qiangdavidliu** [comment]
Thanks @spencergibb . Changes looks clean.

**Comment by spencergibb** [comment]
@qiangdavidliu or @tbak any idea of when you might have a release?

**Comment by qiangdavidliu** [comment]
@spencergibb we are looking at doing an rc release today. Since many of the dependencies has been updated, we will make the full release once the rc from today is confirmed to be stable.

**Comment by spencergibb** [comment]
@qiangdavidliu thanks for the update!  Our team really appreciates it.

**Pull request merged**